include/rmm/device_uvector.hpp

```diff
@@ -0,0 +1,262 @@
+ *
+ * // Initializes all elements to 0
+ * thrust::uninitialized_fill(thrust::device, uv.begin(), uv.end(), 0);
+ * @endcode
+ *
+ * Avoiding default initialization improves performance by eliminating the kernel launch required to
+ * default initialize the elements. This initialization is often unnecessary, e.g., when the vector
+ * is created to hold some output from some operation.
+ *
+ * However, this restricts the element type `T` to only trivially copyable types. In short,
+ * trivially copyable types can be safely copied with `memcpy`. For more information, see
+ * https://en.cppreference.com/w/cpp/types/is_trivially_copyable.
+ *
+ * Another key difference over `thrust::device_vector` is that all operations that invoke
+ * allocation, kernels, or memcpys take a CUDA stream parameter to indicate on which stream the
+ * operation will be performed.
+ *
+ * @tparam T Trivially copyable element type
+ */
+template <typename T>
+class device_uvector {
+  static_assert(std::is_trivially_copyable<T>::value,
+                "device_uvector only supports types that are trivially copyable.");
+
+ public:
+  using value_type      = T;
+  using size_type       = std::size_t;
+  using reference       = value_type&;
+  using const_reference = value_type const&;
+  using pointer         = value_type*;
+  using const_pointer   = value_type const*;
+  using iterator        = pointer;
+  using const_iterator  = const_pointer;
+
+  device_uvector()                      = default;
+  ~device_uvector()                     = default;
+  device_uvector(device_uvector const&) = default;
+  device_uvector(device_uvector&&)      = default;
+  device_uvector& operator=(device_uvector const&) = default;
+  device_uvector& operator=(device_uvector&&) = default;
+
+  /**
+   * @brief Construct a new `device_uvector` with sufficient uninitialized storage for `size`
+   * elements.
+   *
+   * Elements are uninitialized. Reading an element before it is initialized results in undefined
+   * behavior.
+   *
+   * @param size The number of elements to allocate storage for
+   * @param stream The stream on which to perform the allocation
+   * @param mr The resource used to allocate the device storage
+   */
+  explicit device_uvector(std::size_t size,
+                          cudaStream_t stream                 = 0,
+                          rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource())
+    : _storage{elements_to_bytes(size), stream, mr}
+  {
+  }
+
+  /**
+   * @brief Construct a new device_uvector by deep copying the contents of another `device_uvector`.
+   *
+   * Elements are copied as if it were performed by `memcpy`, i.e., `T`'s copy constructor is not
+   * invoked.
+   *
+   * @param other The vector to copy from
+   * @param stream The stream on which to perform the copy
+   * @param mr The resource used to allocate device memory for the new vector
+   */
+  explicit device_uvector(device_uvector const& other,
```

Comment:
```suggestion
  explicit device_uvector(device_uvector<T> const& other,
```

What on earth is this!? How does this even compile???

Reply:
Man you are a harsh reviewer. :)

Reply:
Apparently within a templated class `C`, the template arguments for `C` are assumed to be the same as the containing instantiation, unless explicitly specified? :man_shrugging:

I went looking for details but came up short.

Also interesting: In addition to the a class template, a constructor may also be templated, but since `Thing<int><float>(...)` isn't a valid syntax, the constructors type parameters _must_ be deduced.

Reply:
In other words, this seems to be perfectly legitimate:
```
template<typename T>
struct X
{
    X(){}
    X(X const& other) { }
    void foo(X other) { }
    X* child;
};
```

Reply:
It makes sense. I guess I was just never explicit aware of this.